File: Cores/ExtremumScalaCore/src/main/scala/OSOL/Extremum/Core/Scala/CodeFeatures/FunctionParser/TreeFunctions/DoubleTreeFunction.scala
package OSOL.Extremum.Core.Scala.CodeFeatures.FunctionParser.TreeFunctions

import OSOL.Extremum.Core.Scala.CodeFeatures.FunctionParser.Parser
import OSOL.Extremum.Core.Scala.CodeFeatures.FunctionParser.Trees.Tree
import OSOL.Extremum.Core.Scala.Vectors.RealVector

class DoubleTreeFunction(t: Tree[Double]) {
  def apply(v: RealVector): Double = t.calculate(v)
}

object DoubleTreeFunction {
  def apply(str: String): DoubleTreeFunction = new DoubleTreeFunction(Parser.parseToDoubleTree(str))
}

File: Cores/OSOL.Extremum.Core.DotNet/CodeFeatures/FunctionParser/TreeFunctions/DoubleTreeFunction.cs
using System.Runtime.CompilerServices;
using OSOL.Extremum.Core.DotNet.CodeFeatures.FunctionParser.Trees;
using OSOL.Extremum.Core.DotNet.Vectors;

namespace OSOL.Extremum.Core.DotNet.CodeFeatures.FunctionParser.TreeFunctions
{
    public class DoubleTreeFunction
    {
        public Tree<double> Tree;
        
        public DoubleTreeFunction(Tree<double> tree)
        {
            this.Tree = tree;
        }

        public DoubleTreeFunction(string str) : this(Parser.BuildTreeD(Parser.ParseString(str)))
        {
            
        }

        public double Calculate(RealVector v) => Tree.Calculate(v);
    }
    
}